server/src/test/java/org/opensearch/index/query/BoolQueryBuilderTests.java

```diff
@@ -456,4 +456,14 @@ public void testMustRewrite() throws IOException {
         IllegalStateException e = expectThrows(IllegalStateException.class, () -> boolQuery.toQuery(context));
         assertEquals("Rewrite first", e.getMessage());
     }
+
+    public void testVisit() {
+        BoolQueryBuilder boolQueryBuilder = new BoolQueryBuilder();
+        boolQueryBuilder.should(new BoolQueryBuilder());
+        boolQueryBuilder.must(new BoolQueryBuilder());
+        boolQueryBuilder.mustNot(new BoolQueryBuilder());
+        boolQueryBuilder.filter(new BoolQueryBuilder());
```

Comment:
I would give each of these its own `TermQueryBuilder`, like

```
boolQueryBuilder.should(new TermQueryBuilder(TEXT_FIELD_NAME, "should");
boolQueryBuilder.must(new TermQueryBuilder(TEXT_FIELD_NAME, "must1");
boolQueryBuilder.must(new TermQueryBuilder(TEXT_FIELD_NAME, "must2"); // Add a second one to confirm that they both get visited
boolQueryBuilder.mustNot(new TermQueryBuilder(TEXT_FIELD_NAME, "mustNot");
boolQueryBuilder.filter(new TermQueryBuilder(TEXT_FIELD_NAME, "filter");
```

Then if you collect all the visited QueryBuilders, you can check that they are all there.